activestorage/lib/active_storage/attached/macros.rb

```diff
@@ -34,19 +34,36 @@ def #{name}
         end
 
         def #{name}=(attachable)
-          #{name}.attach(attachable)
+          @active_storage_attached_#{name}_change =
+            if attachable.nil?
+              ActiveStorage::Attached::Changes::DeleteOne.new("#{name}", self)
+            else
+              ActiveStorage::Attached::Changes::CreateOne.new("#{name}", self, attachable)
+            end
+        end
+
+        def #{name}_change
+          @active_storage_attached_#{name}_change
+        end
+
+        def clear_#{name}_change
+          @active_storage_attached_#{name}_change = nil
         end
       CODE
 
-      has_one :"#{name}_attachment", -> { where(name: name) }, class_name: "ActiveStorage::Attachment", as: :record, inverse_of: :record, dependent: false
+      has_one :"#{name}_attachment", -> { where(name: name) }, class_name: "ActiveStorage::Attachment", as: :record, inverse_of: :record, dependent: :destroy
       has_one :"#{name}_blob", through: :"#{name}_attachment", class_name: "ActiveStorage::Blob", source: :blob
 
       scope :"with_attached_#{name}", -> { includes("#{name}_attachment": :blob) }
 
-      if dependent == :purge_later
-        after_destroy_commit { public_send(name).purge_later }
-      else
-        before_destroy { public_send(name).detach }
+      after_save { public_send("#{name}_change")&.save }
+
+      after_commit(on: [:create, :update]) do
+        begin
+          public_send("#{name}_change").try(:upload)
```

Comment:
Why not use `&.upload` like above?

Reply:
The `Delete*` changes don’t respond to `upload`. They could implement `upload` and do nothing, and I had them do that originally, but it seemed odd to me.